ledger/src/blockstore_metrics.rs

```diff
@@ -516,7 +512,43 @@ pub(crate) fn report_rocksdb_write_perf(metric_header: &'static str) {
 /// A struct that holds the current status of RocksDB perf sampling.
 pub struct PerfSamplingStatus {
     // The number of RocksDB operations since the last perf sample.
-    pub(crate) op_count: AtomicUsize,
+    op_count: AtomicUsize,
+    // The timestamp of the latest operation with perf stats collection.
+    last_sample_time_ms: AtomicU64,
+}
+
+impl PerfSamplingStatus {
+    fn should_sample(&self, sample_count_interval: usize) -> bool {
+        if sample_count_interval == 0 {
+            return false;
+        }
+
+        // Rate-limiting based on the number of samples.
+        if self.op_count.fetch_add(1, Ordering::Relaxed) < sample_count_interval {
+            return false;
+        }
+        self.op_count.store(0, Ordering::Relaxed);
+
+        // Rate-limiting based on the time duration.
+        let current_time_ms = timestamp();
+        let old_time_ms = self.last_sample_time_ms.load(Ordering::Relaxed);
+        if old_time_ms + (PERF_SAMPLING_MIN_DURATION.as_millis() as u64) < current_time_ms {
+            return false;
+        }
+
+        // If the `last_sample_time_ms` has a different value than `old_time_ms`,
+        // it means some other thread has performed the sampling and updated
+        // the last sample time.  In this case, the current thread will skip
+        // the current sample.
+        self.last_sample_time_ms
+            .compare_exchange_weak(
```

Comment:
If you don't want to `loop` here then this should be `compare_exchange` instead of `compare_exchange_weak`.

Reply:
Quote from the [compare_exchange_weak](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html#method.compare_exchange_weak) doc:

> Unlike [AtomicUsize::compare_exchange](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html#method.compare_exchange), this function is allowed to spuriously fail even when the comparison succeeds, which can result in more efficient code on some platforms.

Does it mean if we use `compare_exchange_weak` here, we are more performant with slightly fewer samples in case of spurious failure.  Am I right at this part?

Reply:
yeah, i  think so. though i don't know how often this might fail spuriously

Reply:
Offline discussed with @behzadnouri about the use of a loop, and here's the summary:

If it happens often, then:
* with looping, it affects the read/write performance more.
* without looping, we collect fewer perf samples.

If it doesn't happen often, then both ways have similar effects.

As we weigh more on the performance than the number of perf samples, we decide to go without using a loop here.